### PR TITLE
Allow MPT models to return attention weights

### PR DESCRIPTION
**Pull request by lorabit110**
Previously, output_attentions is not propagated to the attention layer. Even when it's set to True, no attention weights were returned.

### CODE REVIEW

**Review by dakinggg — COMMENTED** [review]
Could you please add a simple unit test for this?

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -434,6 +434,7 @@ def forward(
                 attn_bias=attn_bias,
                 attention_mask=attention_mask,
                 is_causal=self.is_causal,
+                output_attentions=output_attentions == True,
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:437]:
```suggestion
                output_attentions=bool(output_attentions),
```
any reason this version doesn't work? the type is `Optional[bool]`, so should either be `None` or a `bool` and `bool(None)` is `False`

**Comment by vchiley** [comment]
[This error](https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/mpt/modeling_mpt.py#L321C22-L321C22) should be noted if you plan to use it.

**Comment by lorabit110** [comment]
> Could you please add a simple unit test for this?

Updated an existing unit test to check attention.

**Comment by lorabit110** [comment]
> [This error](https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/models/mpt/modeling_mpt.py#L321C22-L321C22) should be noted if you plan to use it.

Yeah. I am aware of that flash attention won't return attention weights.

File: tests/test_model.py [hunk 2]
```diff
@@ -1341,6 +1341,7 @@ def test_forward_with_output_attentions_and_output_hidden_states(
 
         if output_attentions:
             assert len(outputs.attentions) == n_layers
+            assert outputs.attentions[0] is not None
```
Comment (dakinggg) [inline tests/test_model.py:1344]:
```suggestion
            assert all(attn.shape == (1, 4, 2048, 2048) for attn in outputs.attentions)
```

**Review by dakinggg — APPROVED** [review]
One comment to make sure we're returning the right shape of stuff, but otherwise lgtm! Thanks for the PR!

File: tests/test_model.py [hunk 3]
```diff
@@ -1341,6 +1341,7 @@ def test_forward_with_output_attentions_and_output_hidden_states(
 
         if output_attentions:
             assert len(outputs.attentions) == n_layers
+            assert all(attn.shape == (1, 4, 2048, 2048) for attn in outputs.attentions)
```
Comment (dakinggg) [inline tests/test_model.py:1344]:
```suggestion
            assert all(attn.shape == (1, 4, 3, 3) for attn in outputs.attentions)
```

**Comment by dakinggg** [comment]
@lorabit110 could you run `precommit run --all-files` locally? Thanks!

File: tests/test_model.py [hunk 4]
```diff
@@ -1341,6 +1341,7 @@ def test_forward_with_output_attentions_and_output_hidden_states(
 
         if output_attentions:
             assert len(outputs.attentions) == n_layers
+            assert all(attn.shape == (1, 4, 3, 3) for attn in outputs.attentions)
```
Comment (dakinggg) [inline tests/test_model.py:1344]:
```suggestion
            assert all(
                attn.shape == (1, 4, 3, 3) for attn in outputs.attentions)
```

**Pull request merged**